### ISSUE TITLE
epic: sequence arc — LSTM/GRU, char-RNN and seq2seq translation with attention (non-transformer sequence modelling)

### ISSUE THREAD

**Issue by bkc39** [bkc39/rktorch#153]
Build the recurrent side of the sequence story: LSTM and GRU layers, token
sampling primitives, and two capstones, a character-level LSTM language model
and an encoder-decoder translator with attention. The transformer arc (#22,
`06-gpt.rkt`) is the only sequence model in the library today; this arc
mirrors ocaml-torch's `char_rnn` and `translation` examples so the library
covers non-transformer sequence modelling as well.

## Why translation
The PyTorch seq2seq tutorial's English-French pairs
(`download.pytorch.org/tutorial/data.zip`, `eng-fra.txt`, ~135k lines,
filtered to short sentences → ~10k pairs) is a plain UTF-8 tab-separated file,
parseable in pure Racket like the Gutenberg text pipeline in
`torch/data/text.rkt`. It exercises everything a recurrent stack needs at
once: word-level vocabularies for two languages, an encoder that consumes a
sequence into a state, a decoder that unrolls one step at a time with teacher
forcing, Bahdanau attention over encoder states, and greedy decoding with
`topk`. ocaml-torch's demo reaches usable translations in under an hour on a
GPU; the 3090 Ti does it in minutes. The char-RNN capstone comes first because
it needs only the LSTM and the existing text pipeline.

## Legs
- **Leg 0 — op tranche**: `lstm.input` and `gru.input` (the cudnn-backed
  multi-layer ATen entry points; batch-first, dropout between layers,
  bidirectional flag), `multinomial`, `topk`, `sort`, `argsort`, `bmm`
  surfaced if `matmul` batching proves insufficient, `nll_loss` promoted to
  the Racket loss surface. Codegen allowlist tranche, recipes, goldens; the
  ATen RNN signatures take a tensor list of flat weights, which is a new
  codegen kind (tensor-array) or a hand-written shim per the AGENTS.md rule.
- **Leg 1 — layers**: `LSTM` and `GRU` in `torch/nn/` with PyTorch's
  parameter naming (`weight_ih_l0`, `weight_hh_l0`, `bias_ih_l0`,
  `bias_hh_l0`, reverse suffixes when bidirectional) so state dicts load
  from `nn.LSTM`; forward returns output and final state as values;
  `#:num-layers`, `#:bidirectional?`, `#:dropout`; init parity under
  `manual_seed`. Gradient-norm clipping (`clip-grad-norm!`) beside the
  optimizers, since both capstones use it.
- **Leg 2 — char-RNN**: `12-char-rnn.rkt` literate example, one-hot or
  embedded characters over the Heart of Darkness pipeline already in
  `torch/data/text.rkt`, LSTM + linear head, Adam + clipping, sampling by
  `softmax` + `multinomial` with temperature; Python twin with fixture-scale
  loss parity; a generated excerpt in the issue.
- **Leg 3 — translation dataset**: `torch/data/translation.rkt` (or the
  modality-first home from #88): cached download, unicode normalisation and
  the tutorial's filtering, two `vocab`s, encode/decode, a committed fixture
  of a few hundred pairs; tests.
- **Leg 4 — seq2seq with attention**: `13-translation.rkt`: GRU encoder,
  GRU decoder with Bahdanau attention (`cat`, `bmm`/`matmul`, `softmax`),
  teacher forcing at p = 0.5 drawn from the bound generator, NLL over
  `log-softmax`, greedy decode with `topk` k = 1; Python twin; held-out
  evaluation by token edit distance (the `edit-distance` from
  `torch/audio/functional.rkt`, generalised or moved) with a handful of
  printed translations in the issue.

Quality bar: parity oracles per leg as in #84; the char-RNN sample and ten
held-out translations recorded on the issue; BLEU is out of scope.

Shared-primitive ownership with the classic vision arc (#152): this arc
owns gradient clipping, `multinomial`/`topk`/`sort` and the recurrent layers;
the vision arc owns the optimizer family, schedulers, half precision and the
vision losses. Both branch from master independently; whoever lands second
rebases over the allowlist, `foreign.rkt` provides, `nn.rkt` provides and the
AGENTS.md rosters.

Transformer predecessor: #22 (`06-gpt.rkt`); audio counterpart: #83.

**Comment by bkc39** [comment]
## Reservations for running #152 and #153 in parallel

Both arcs branch from master independently, in separate worktrees, and the
one that merges second rebases. These reservations keep the rebase mechanical.

**Numbering**
- allowlist and gtest tranches: #152 = tranche 6 (`generated_tranche6_test.cpp`),
  #153 = tranche 7 (`generated_tranche7_test.cpp`); each inserts its own line
  in `cpp/CMakeLists.txt`'s sorted list.
- examples: #152 = `09-resnet`, `10-dcgan`, `11-vae`; #153 = `12-char-rnn`,
  `13-translation`; each adds its own `examples/test/info.rkt` timeout entry.

**File ownership for the duration**
- `torch/nn/optim.rkt` belongs to #152 (lr hook on `gen:optimizer`, momentum,
  nesterov, weight decay, rmsprop) with schedulers in `torch/nn/scheduler.rkt`.
- #153 puts `clip-grad-norm!` in a new `torch/nn/clip.rkt`, not in optim.rkt.
- `codegen/ir.py`, `codegen/emit_*.py`, `torch/foreign/define-generated.rkt`
  belong to #153 (multi-output returns, #154). #152's ops need no
  generator change.
- `torch/nn/loss.rkt` is append-only for both (#152: bce-with-logits, huber,
  l1; #153: nll-loss).
- half precision (dtype enum, `to-dtype`, safetensors tags, `with-autocast`)
  belongs to #152.
- Shared hot spots both edit, rebase by hand: `codegen/allowlist.txt`,
  `cpp/tests/torchrkt/c_api_compile_test.c`, `torch/foreign.rkt` and
  `torch/nn.rkt` provides, `torch/tests/generated-parity-test.rkt` recipes,
  AGENTS.md rosters.

**Refactor freeze while both arcs are open**: #141 (layers home), #117
(define-layer syntax), #112 (functional facade), #91 and #136 (import UX).
They would conflict with every layer either arc adds.

**GPU etiquette on the single 3090 Ti** (shared with the #145 memory session):
- never run the `.#ci` and `.#cuda` shells concurrently in one worktree (each
  restages the native shim; the other sees "CUDA support is not compiled").
- run CUDA suites only when `nvidia-smi` shows under 2 GB used; the
  wait-then-run pattern from #138 is the reference.
- announce full training runs on the epic with start time and expected
  duration; the sequence capstones are minutes, the ResNet run is tens of
  minutes, so the vision session yields to the other two.
- #147 (#145) should merge before the ResNet full run if it is ready; no leg
  0 waits for it.

**Comment by bkc39** [comment]
Leg 0 precursor: #154 (multi-output returns in the generator) is the sequence session's first PR; topk and sort are its first customers, then lstm.input and gru.input.

**Comment by bkc39** [comment]
Sequence session status (worktree `sequence-153`):

- #156 closes #154: multi-Tensor returns in the generator (`Op.returns`, `alloc_results`, `#:returns N`, `tensor-allocator/outputs`), `topk` and `sort` as first customers. Draft, local gates green, CI green bar macOS still building.
- #157 is leg 0 (tranche 7), stacked on #156: `optional-generator` IR kind, `lstm.input`, `gru.input`, `multinomial`, `argsort`; `topk` / `sort` / `argsort` / `multinomial` promoted to `torch`, `nll-loss` to `torch/nn`. Local gates and the CUDA-guarded suites green.

One thing the vision session should know: `sort` now shadow-dispatches on `torch`, so the literate examples' `(for-label (except-in racket/base ...))` list gains `sort`. `09`-`11` need the same word when #152 rebases over #157.

GPU use so far: two sub-minute test runs. Leg 1 (LSTM/GRU layers, `clip-grad-norm!` in `torch/nn/clip.rkt`) is next, stacked on #157.

**Comment by bkc39** [comment]
GPU notice (sequence session): starting the leg 2 char-RNN full run now (19:29Z), two-layer 256-wide LSTM on the whole novella, 30 epochs, expected under 10 minutes and a few hundred MB of device memory.

**Comment by bkc39** [comment]
GPU notice (sequence session): leg 2's char-RNN runs finished (25 s and ~1 min). Starting the leg 4 translation full run now (20:29Z): GRU seq2seq with attention on the 11,445 tutorial pairs, 60 epochs, expected about 6 minutes and under 1 GB of device memory.

**Comment by bkc39** [comment]
Sequence arc: every leg is written, locally gated (`nix flake check`, PyTorch twins, CUDA-guarded suites on the 3090 Ti) and up as a linear stack, each PR based on the one below. GPU runs are finished; the card is free.

| | PR | branch | state |
|---|---|---|---|
| #154 multi-output codegen | #156 | `seq/rnn-153` | ready for review, CI 6/6; Codex round: one finding fixed (b3b29bc), one declined with reasons |
| leg 0, tranche 7 | #157 | `seq/tranche7-153` | draft, CI 5/5 |
| leg 1, LSTM/GRU + `clip-grad-norm!` | #160 | `seq/layers-153` | draft |
| leg 2, `12-char-rnn` | #163 | `seq/char-rnn-153` | draft |
| leg 3, translation pairs | #164 | `seq/translation-data-153` | draft |
| leg 4, `13-translation` | #166 | `seq/translation-153` | draft |

Merging is bottom-up: after #156 squash-merges, #157 gets master merged in, retargets to master and goes ready, and so on. Fixes to a lower PR are carried upward by merging that branch into the ones above (no force-pushes).

**Two things for the owner**
- #164 commits a 318-line excerpt of the tutorial's eng-fra pairs, which are Tatoeba's, **CC BY 2.0 FR**. There is a `NOTICE` beside it; `info.rkt`'s license field is untouched and is your call.
- For #152: `sort` now shadow-dispatches on `torch`, so the examples' `(for-label (except-in racket/base ...))` list gains `sort`; `09`-`11` need the same word on rebase. `take` is also a torch name, which bit me once in an example runner.

**Char-RNN sample** (two-layer 256-wide LSTM, whole novella, 80 epochs ≈ 1 min, loss 3.06 → 0.98, temperature 0.8, prompt `The `):

> The creeked mesided in from pryjist. ‘Steer with a broad cabre ordaring the open conspares or sometimes he could not see the man who is a mark flying overhapperited in the dark in the doorwith steaming stars with blue thought, so was lying up. His farthing shook becking, complete. I did not appear to the forest, by about the purpose. It seemed to say the coast of these matters I could be one humbered in a clump of yells, but a bend outward to me, I could see a dead that in these distance. He was every engine—I am not surprised his became vision at the black fellows to leave!

**Translation** (GRU encoder, Bahdanau-attention GRU decoder, hidden 128, 10,301 training pairs, 60 epochs = 4 min 59 s): held-out **token error rate 0.296** over 1,144 unseen pairs. First ten held-out pairs:

```
> tu m enerves                              = you re annoying me              < you re avoiding me
> ton temps est ecoule                      = you re out of time              < we re out of time
> je fais la fete                           = i m celebrating                 < i m doing the food
> je n ai vraiment pas aussi faim que ca    = i m really not all that hungry  < i m really not as hungry as
> je suis employe de banque                 = i m a bank clerk                < i m a bank a bank
> ils s ennuient                            = they are bored                  < they are bored
> je ne suis pas inquiet pour tom           = i m not worried about tom       < i m not worried about tom
> il est un peu rugueux                     = he s a bit rough around the edges  < he s a bit up now
> ce sont tous les miens                    = they re all mine                < they re all mine
> j arrive                                  = i am coming                     < i m coming
```

Not done, by design: `bmm` (batched `matmul` sufficed), moving `edit-distance` (it already compares arbitrary lists), unbatched/packed/projected RNN inputs.

**Comment by bkc39** [comment]
Reviewer-thread status for the sequence stack (in answer to the request relayed by the auxiliary session):

- #156 (b3b29bc): 2 Codex threads, **both resolved**. The out-slot one is fixed in b3b29bc; the accounting-inside-the-atomic-section one is declined with reasons on the thread (a break there unwinds the wrapper, so the handles never reach the caller and their finalizers free them; it is also the single-output wrap's shape since #37, and changing both would land in `memory.rkt` beside #147's edits).
- #157 584528c, #160 d82c54f, #163 3aa4f15, #164 bd85983, #166 65eda2b: 0 threads, CI green, `CLEAN`. They were drafts, and Codex only reviews on ready-for-review, so I have marked all five ready to get their Codex rounds done before your manual review. I will clear whatever they raise (fix or reasoned decline, then resolve; one batched push per round, carried up the stack by merging; no force-pushes, no merges of PRs).
- #167 is the vision session's, not part of this stack.

Order is strict: #156 → #157 → #160 → #163 → #164 → #166, each based on the branch below. If you would rather have one merge cycle than six, #166 is the tip and contains everything: retarget it to master and close the five below. After any allowlist PR merges (#156, #157, #158, #165) I will merge `origin/master` and re-run `nix run .#codegen` rather than hand-resolve generated files.

**Comment by bkc39** [comment]
GPU notice (sequence session): re-running the leg 4 translation training now (22:46Z) after a review fix to the encoder state; 60 epochs, about 5 minutes, under 1 GB.

**Comment by bkc39** [comment]
Bot rounds are done on the sequence stack: #156 b3b29bc, #157 390efe4, #160 12030ef, #163 b57b117, #164 0bb3d27, #166 97b6041 are all CLEAN, CI green on ubuntu and macOS, 20 Codex threads across them all resolved (15 fixed, 5 declined with reasons on the thread). Two changes from the rounds worth knowing at review: LSTM/GRU forget their cudnn flattening on every move (their own `prop:to`), and the translation encoder hands the decoder the state at `<eos>` rather than after the padding (retrained: held-out token error rate 0.304, within noise of the 0.296 before; the PR body has the new numbers). Ready for your manual review, bottom-up.

**Comment by bkc39** [comment]
All five legs are merged:

| Leg | PR | Squash |
|---|---|---|
| multi-output returns in the generator; `topk`, `sort` | #156 | 288e32f |
| tranche 7: `lstm.input`, `gru.input`, `multinomial`, `argsort`, `nll-loss` surfaced | #157 | b5e5d41 |
| 1 — `LSTM`, `GRU` as `define-layer` forms, `clip-grad-norm!` | #160 | ce04fee |
| 2 — `12-char-rnn`, a character-level LSTM | #163 | ea1abe1 |
| 3 — eng-fra translation pairs, CC-BY-2.0-FR declared | #164 | 342a5ea |
| 4 — `13-translation`, GRU seq2seq with Bahdanau attention | #166 | 4bc4973 |

Measured on the 3090 Ti after the merges: the char-RNN reaches loss 1.37 in 13 s and writes better text than the GPT example at the same parameter count; the seq2seq reproduces its held-out token error rate of 0.304 and, on the full 105,692-pair corpus the same download yields with `#:prefixes #f`, reaches 0.336 at 20 epochs.

Along the way `define-layer` grew `parameters-by-key`, rest-argument `#:forward` and `#:on-move`, and `gen:layer` grew `layer-own-tensors`.

Follow-ups filed from the reviews rather than folded in: #178 forward-contract validation of recurrent state shapes (now unblocked by #152's `#:forward` contracts), #184 flat-weight invalidation inside `to`, #185 vanilla RNN and a shared base, #187 batched `clip-grad-norm!`, #189 the flat-weight cache under concurrency, #193 shared env-override parsing in the example harnesses, #195 one cached-download helper for five copies, and #91/#136 for the example import boilerplate.
